File: game/src/world/islandGen.ts
/**
 * Floating voxel island — port of Design/blender_scripts/islands.py `generate_island`.
 *
 * Blender is z-up; here the horizontal plane is (x, z) and height is y.
 * Origin = island centre, y = 0 is the lowest grass layer, the island hangs below.
 */
import { BLOCK } from './palette'
import { makeRng, PerlinNoise, type Rng } from './noise'
import type { World } from './chunkStore'

export interface IslandParams {
  size: number
  seed: number
  maxHeight: number
  depth: number
  padRadius: number
  lake: boolean
  treeDensity: number
}

/** Island_Large from islands.py ISLANDS table. */
export const ISLAND_LARGE: IslandParams = {
  size: 56, seed: 11, maxHeight: 9, depth: 16, padRadius: 8, lake: true, treeDensity: 0.022,
}

export const ISLAND_MEDIUM: IslandParams = {
  size: 32, seed: 7, maxHeight: 6, depth: 11, padRadius: 5, lake: true, treeDensity: 0.022,
}

export interface IslandInfo {
  /** feet position on top of the build pad */
  spawn: { x: number; y: number; z: number }
  voxelCount: number
}

const colKey = (x: number, z: number): number => (x + 512) * 1024 + (z + 512)

/** 2-D fractal noise in roughly -1..1 (`n2` in islands.py). */
function makeN2(perlin: PerlinNoise) {
  return (x: number, z: number, seed: number, freq: number): number => {
    const vx = x * freq
    const vz = z * freq
    const vs = seed * 7.31
    return perlin.noise(vx, vz, vs) + 0.5 * perlin.noise(vx * 2.1 + 3.3, vz * 2.1 + 1.7, vs * 2.1)
  }
}

function addTree(world: World, x: number, y: number, z: number, rng: Rng): void {
  const height = rng.randint(4, 6)
  for (let dy = 0; dy < height; dy++) world.setBlock(x, y + dy, z, BLOCK.log)
  const top = y + height
  for (let dy = -2; dy < 2; dy++) {
    const r = dy < 0 ? 2 : 1
    for (let dx = -r; dx <= r; dx++) {
      for (let dz = -r; dz <= r; dz++) {
        if (Math.abs(dx) === r && Math.abs(dz) === r && dy !== -1) continue
        world.setBlockIfAir(x + dx, top + dy, z + dz, BLOCK.leaves)
      }
    }
  }
  world.setBlock(x, top + 2, z, BLOCK.leaves)
}

export function generateIsland(world: World, p: IslandParams): IslandInfo {
  const { size, seed, maxHeight, depth, padRadius, lake, treeDensity } = p
  const rng = makeRng(seed)
  const n2 = makeN2(new PerlinNoise(seed))
  const R = size / 2
  const Ri = Math.trunc(R)

  // ---- heightmap + island mask
  const heights = new Map<number, { x: number; z: number; h: number; dep: number }>()
  for (let x = -Ri - 1; x <= Ri + 1; x++) {
    for (let z = -Ri - 1; z <= Ri + 1; z++) {
      const d = Math.hypot(x, z) / R
      const mask = 1 - d + 0.3 * n2(x, z, seed, 0.09)
      if (mask <= 0.12) continue
      let h = mask * maxHeight + 2 * n2(x, z, seed + 1, 0.16)
      h = mask > 0.5 ? Math.round(h / 2) * 2 : Math.round(h) // terraces inland
      h = Math.max(0, Math.min(maxHeight + 2, h))
      if (padRadius && Math.hypot(x, z) <= padRadius) h = 2
      let dep = Math.trunc(mask * depth + 2 * n2(x, z, seed + 2, 0.13) * mask)
      dep = Math.max(1, dep)
      heights.set(colKey(x, z), { x, z, h, dep })
    }
  }

  // ---- lake: pick a spot away from the pad
  const lakeCells = new Set<number>()
  let lakeLevel = 0
  if (lake) {
    const lr = Math.max(2, Math.trunc(size / 9))
    // islands.py samples a square and only keeps the lake *centre* off the pad, which
    // can leave the lake (radius + noise + sand shore) cutting into the pad. Sample a
    // ring instead so the whole lake always clears the pad.
    const minLakeDist = padRadius + lr + 3
    const maxLakeDist = Math.max(minLakeDist, R * 0.7)
    const angle = rng.random() * Math.PI * 2
    const dist = minLakeDist + rng.random() * (maxLakeDist - minLakeDist)
    const lx = Math.round(Math.cos(angle) * dist)
    const lz = Math.round(Math.sin(angle) * dist)
    for (const { x, z } of heights.values()) {
      if (Math.hypot(x - lx, z - lz) <= lr + 0.6 * n2(x, z, seed + 3, 0.3)) lakeCells.add(colKey(x, z))
    }
    if (lakeCells.size) {
      lakeLevel = Infinity
      for (const k of lakeCells) lakeLevel = Math.min(lakeLevel, heights.get(k)!.h)
    }
  }

  // ---- fill columns
  let voxelCount = 0
  for (const { x, z, h, dep } of heights.values()) {
    if (lakeCells.has(colKey(x, z))) {
      const floor = lakeLevel - 2
      for (let y = -dep; y < floor; y++) { world.setBlock(x, y, z, BLOCK.stone); voxelCount++ }
      world.setBlock(x, floor, z, BLOCK.sand)
      voxelCount++
      for (let y = floor + 1; y <= lakeLevel; y++) { world.setBlock(x, y, z, BLOCK.water); voxelCount++ }
      continue
    }
    let nearLake = false
    for (let dx = -1; dx <= 1 && !nearLake; dx++) {
      for (let dz = -1; dz <= 1; dz++) {
        if (lakeCells.has(colKey(x + dx, z + dz))) { nearLake = true; break }
      }
    }
    for (let y = -dep; y <= h; y++) {
      let kind: number
      if (y === h) {
        kind = nearLake && h <= lakeLevel + 1 ? BLOCK.sand : BLOCK.grass
      } else if (y >= h - 2) {
        kind = BLOCK.dirt
      } else {
        kind = BLOCK.stone
        const r = rng.random()
        if (r < 0.03) kind = BLOCK.ore_coal
        else if (r < 0.045) kind = BLOCK.ore_iron
      }
      world.setBlock(x, y, z, kind)
      voxelCount++
    }
    // snow on the highest peaks of big islands
    if (h >= maxHeight + 1 && size >= 40) world.setBlock(x, h, z, BLOCK.snow)
  }

  // ---- trees on grass, not on the pad
  const grassCells: { x: number; y: number; z: number }[] = []
  for (const { x, z, h } of heights.values()) {
    if (lakeCells.has(colKey(x, z))) continue
    if (world.getBlock(x, h, z) !== BLOCK.grass) continue
    const dist = Math.hypot(x, z)
    if (dist > padRadius + 1 && dist < R * 0.85) grassCells.push({ x, y: h, z })
  }
  const shuffled = rng.shuffle(grassCells)
  const placed: { x: number; z: number }[] = []
  const target = Math.trunc(grassCells.length * treeDensity)
  for (const { x, y, z } of shuffled) {
    if (placed.length >= target) break
    if (placed.every(p => Math.abs(x - p.x) > 3 || Math.abs(z - p.z) > 3)) {
      addTree(world, x, y + 1, z, rng)
      placed.push({ x, z })
    }
  }

  return { spawn: { x: 0.5, y: 3, z: 0.5 }, voxelCount }
}

File: client/lib/world/island_gen.dart
/// Floating voxel island — twin of `server/resources/js/world/islandGen.ts`
/// (itself a port of Design/blender_scripts/islands.py `generate_island`).
///
/// Origin = island centre, y = 0 is the lowest grass layer, the island hangs below.
library;

import 'js_math.dart';
import 'noise.dart';
import 'palette.dart';

/// where generated voxels go: a template that is stamped into chunks later
abstract interface class VoxelSink {
  int getBlock(int x, int y, int z);
  void setBlock(int x, int y, int z, int id);
  void setBlockIfAir(int x, int y, int z, int id);
}

final class IslandParams {
  const IslandParams({
    required this.size,
    required this.seed,
    required this.maxHeight,
    required this.depth,
    required this.padRadius,
    required this.lake,
    required this.treeDensity,
  });

  final int size;
  final int seed;
  final int maxHeight;
  final int depth;
  final int padRadius;
  final bool lake;
  final double treeDensity;
}

/// Island_Large from islands.py — the legacy island over the spawn pad
const IslandParams islandLarge = IslandParams(
  size: 56,
  seed: 11,
  maxHeight: 9,
  depth: 16,
  padRadius: 8,
  lake: true,
  treeDensity: 0.022,
);

const IslandParams islandMedium = IslandParams(
  size: 32,
  seed: 7,
  maxHeight: 6,
  depth: 11,
  padRadius: 5,
  lake: true,
  treeDensity: 0.022,
);

final class IslandInfo {
  const IslandInfo({required this.padHeight, required this.voxelCount});

  /// y of the pad's top block
  final int padHeight;
  final int voxelCount;
}

final class _Column {
  _Column(this.x, this.z, this.h, this.dep);

  final int x;
  final int z;
  int h;
  final int dep;
}

int _colKey(int x, int z) => (x + 512) * 1024 + (z + 512);

void _addTree(VoxelSink world, int x, int y, int z, Rng rng) {
  final height = rng.randint(4, 6);
  for (var dy = 0; dy < height; dy++) {
    world.setBlock(x, y + dy, z, Block.log);
  }
  final top = y + height;
  for (var dy = -2; dy < 2; dy++) {
    final r = dy < 0 ? 2 : 1;
    for (var dx = -r; dx <= r; dx++) {
      for (var dz = -r; dz <= r; dz++) {
        if (dx.abs() == r && dz.abs() == r && dy != -1) continue;
        world.setBlockIfAir(x + dx, top + dy, z + dz, Block.leaves);
      }
    }
  }
  world.setBlock(x, top + 2, z, Block.leaves);
}

IslandInfo generateIsland(VoxelSink world, IslandParams p) {
  final size = p.size;
  final seed = p.seed;
  final maxHeight = p.maxHeight;
  final depth = p.depth;
  final padRadius = p.padRadius;
  final rng = Rng(seed);
  final n2 = fractal2(PerlinNoise(seed));
  final bigR = size / 2;
  final ri = jsTrunc(bigR);
  final seedD = seed.toDouble();

  // ---- heightmap + island mask (insertion order matters: the RNG is consumed in it)
  final heights = <int, _Column>{};
  for (var x = -ri - 1; x <= ri + 1; x++) {
    for (var z = -ri - 1; z <= ri + 1; z++) {
      final xd = x.toDouble();
      final zd = z.toDouble();
      final d = hypot(xd, zd) / bigR;
      final mask = 1 - d + 0.3 * n2(xd, zd, seedD, 0.09);
      if (mask <= 0.12) continue;
      final hRaw = mask * maxHeight + 2 * n2(xd, zd, seedD + 1, 0.16);
      var h = mask > 0.5
          ? jsRound(hRaw / 2) * 2
          : jsRound(hRaw); // terraces inland
      h = h < 0 ? 0 : (h > maxHeight + 2 ? maxHeight + 2 : h);
      var dep = jsTrunc(mask * depth + 2 * n2(xd, zd, seedD + 2, 0.13) * mask);
      dep = dep < 1 ? 1 : dep;
      heights[_colKey(x, z)] = _Column(x, z, h, dep);
    }
  }

  // ---- build pad: flat, at the height of the surrounding terrain
  var padHeight = 2;
  if (padRadius != 0) {
    final ring = <int>[];
    for (final c in heights.values) {
      final dist = hypot(c.x.toDouble(), c.z.toDouble());
      if (dist > padRadius && dist <= padRadius + 3) ring.add(c.h);
    }
    if (ring.isNotEmpty) {
      ring.sort();
      padHeight =
          jsRound(ring[ring.length >> 1] / 2) *
          2; // median, snapped to a terrace
    }
    for (final c in heights.values) {
      if (hypot(c.x.toDouble(), c.z.toDouble()) <= padRadius) c.h = padHeight;
    }
  }

  // ---- lake: pick a spot on a ring that clears the pad
  final lakeCells = <int>{};
  var lakeLevel = 0;
  if (p.lake) {
    final lrRaw = jsTrunc(size / 9);
    final lr = lrRaw > 2 ? lrRaw : 2;
    final minLakeDist = (padRadius + lr + 3).toDouble();
    final maxLakeRaw = bigR * 0.7;
    final maxLakeDist = maxLakeRaw > minLakeDist ? maxLakeRaw : minLakeDist;
    final angle = rng.random() * jsPi * 2;
    final dist = minLakeDist + rng.random() * (maxLakeDist - minLakeDist);
    final lx = jsRound(jsCos(angle) * dist);
    final lz = jsRound(jsSin(angle) * dist);
    for (final c in heights.values) {
      if (hypot((c.x - lx).toDouble(), (c.z - lz).toDouble()) <=
          lr + 0.6 * n2(c.x.toDouble(), c.z.toDouble(), seedD + 3, 0.3)) {
        lakeCells.add(_colKey(c.x, c.z));
      }
    }
    if (lakeCells.isNotEmpty) {
      var lowest = 1 << 30;
      for (final k in lakeCells) {
        final h = heights[k]!.h;
        if (h < lowest) lowest = h;
      }
      lakeLevel = lowest;
    }
  }

  // ---- fill columns
  var voxelCount = 0;
  for (final c in heights.values) {
    final x = c.x;
    final z = c.z;
    final h = c.h;
    if (lakeCells.contains(_colKey(x, z))) {
      final floor = lakeLevel - 2;
      for (var y = -c.dep; y < floor; y++) {
        world.setBlock(x, y, z, Block.stone);
        voxelCount++;
      }
      world.setBlock(x, floor, z, Block.sand);
      voxelCount++;
      for (var y = floor + 1; y <= lakeLevel; y++) {
        world.setBlock(x, y, z, Block.water);
        voxelCount++;
      }
      continue;
    }
    var nearLake = false;
    for (var dx = -1; dx <= 1 && !nearLake; dx++) {
      for (var dz = -1; dz <= 1; dz++) {
        if (lakeCells.contains(_colKey(x + dx, z + dz))) {
          nearLake = true;
          break;
        }
      }
    }
    for (var y = -c.dep; y <= h; y++) {
      int kind;
      if (y == h) {
        kind = nearLake && h <= lakeLevel + 1 ? Block.sand : Block.grass;
      } else if (y >= h - 2) {
        kind = Block.dirt;
      } else {
        kind = Block.stone;
        // one roll, as before: adding copper to the ladder must not shift the island's RNG
        final r = rng.random();
        if (r < 0.03) {
          kind = Block.oreCoal;
        } else if (r < 0.045) {
          kind = Block.oreIron;
        } else if (r < 0.055) {
          kind = Block.oreCopper;
        }
      }
      world.setBlock(x, y, z, kind);
      voxelCount++;
    }
    // snow on the highest peaks of big islands
    if (h >= maxHeight + 1 && size >= 40) world.setBlock(x, h, z, Block.snow);
  }

  // ---- trees on grass, not on the pad
  final grassCells = <(int, int, int)>[];
  for (final c in heights.values) {
    if (lakeCells.contains(_colKey(c.x, c.z))) continue;
    if (world.getBlock(c.x, c.h, c.z) != Block.grass) continue;
    final dist = hypot(c.x.toDouble(), c.z.toDouble());
    // +3: the radius-2 canopy must not overhang the pad
    if (dist > padRadius + 3 && dist < bigR * 0.85) {
      grassCells.add((c.x, c.h, c.z));
    }
  }
  final shuffled = rng.shuffle(grassCells);
  final placed = <(int, int)>[];
  final target = jsTrunc(grassCells.length * p.treeDensity);
  for (final (x, y, z) in shuffled) {
    if (placed.length >= target) break;
    if (placed.every((q) => (x - q.$1).abs() > 3 || (z - q.$2).abs() > 3)) {
      _addTree(world, x, y + 1, z, rng);
      placed.add((x, z));
    }
  }

  return IslandInfo(padHeight: padHeight, voxelCount: voxelCount);
}
